fix lowranceusr4 writer issues

tsteven4 opened the pull request:
This should resolve #375.

In addition, further writer issues with lowranceusr version 4 should be resolved:

- BOMs were erroneously included in internal UTF-16LE strings.
- Timestamps were erroneously offset by 12 hours.
- The written number of mysterious attribute bytes didn't match the written count.

Also, rounding is now used when converting real latitude and longitude numbers to the integer format used by all lowranceusr files.  This makes round trip testing much more likely to succeed.

tsteven4 committed:
fixes for lowraneusr writer for versions >= 4.

use fs_chain_find when looking for FS_LOWRANCEUSR4 data.
and check for nullptr to verify FS_LOWRANCEUSR4 data exists.

tsteven4 committed:
using rounding in lowranceusr lat/lon conversions.

This gives as a chance of usr->gpx->usr matching.

Update gpsbabel generated usr reference files for these
rounding differences.

tsteven4 committed:
fix lowranceusr4 writer issues.

when writing embedded UTF-16LE strings in lowranceusr version >=4
don't inlucde BOM.

don't offset returned values from lowranceusr4_jd_from_timestamp
by 12 hours, thus matching the inverse transformation in
lowranceusr4_get_timestamp.

when writing lowranceusr version >=4 trails output the given number
of mysterious attribute bytes, i.e. 0.

This almost enables a gpx -> usr -> gpx test loop, which is included
but commented in lowranceusr.test.

GPSBabelDeveloper commented on testo.d/lowranceusr.test:
```diff
@@ -68,22 +68,22 @@ rm -f ${TMPDIR}/lowrance*
-# Unfortunately precision issues cause mismatch on Lat/Long conversion, actual test validation commented out
-# bincompare ${REFERENCE}/lowrance.usr  ${TMPDIR}/lowrance1.usr
+gpsbabel -i lowranceusr -f ${TMPDIR}/lowrance1.usr -o lowranceusr,wversion=2 -F ${TMPDIR}/lowrance2.usr
+# At a minimum unkown altitude not making the round trip, actual test validation commented out
```
unknown

tsteven4 replied:
spelling fixed.

GPSBabelDeveloper reviewed (approved):
I'll leave the final approval to the author of the new code here.
Overall, this looks pretty sane. The fs_find_chain stuff really does need a c++ rethink some day. Maybe the format specific stuff becomes subclasses of points, but we have cases (at least in theory) where we could have multiple format specific data attached. Maybe we just add an array of QVariants (std::variant) of name/value pairs and get rid of all the pointer abuse.
I'll think on that a bit, but this seems like a bunch of good fixes to me...

tsteven4 committed:
fixes to enable lowranceusr v4 round trip testing.

use templates for lowranceusr*_find_desc_from_icon_number
and lowranceusr*_find_icon_number_from_desc.

enhance lowranceusr*_find_icon_number_from_desc to
recongnize icon-n where n is an integer as icon number n.
This is necessary for the v4 round trip test.

declare lowranceusr*_find_desc_from_icon_number to be static
to limit visibilty. Drop const return type.

declare lowranceusr4_find_color_from_icon_number_plus_color_index
to be static to limit visiblity.

tsteven4 committed:
fix new lowranceusr memory leak.
